Userland/Libraries/LibWeb/ServiceWorker/Job.cpp

```diff
@@ -122,14 +194,103 @@ static void run_job(JS::VM& vm, JobQueue& job_queue)
+            // 1. Let convertedValue be null.
+            // 2. If equivalentJob’s job type is either register or update, set convertedValue to the result of
+            //    getting the service worker registration object that represents value in equivalentJob’s client.
+            // 3. Else, set convertedValue to value, in equivalentJob’s client's Realm.
+            // 4. Resolve equivalentJob’s job promise with convertedValue.
+            WebIDL::resolve_promise(realm, *equivalent_job->job_promise, value);
+        }));
+    }
+}
+
+// https://w3c.github.io/ServiceWorker/#reject-job-promise-algorithm
+template<typename Error>
+static void reject_job_promise(JS::NonnullGCPtr<Job> job, FlyString message)
+{
+    // 1. If job’s client is not null, queue a task, on job’s client's responsible event loop using the DOM manipulation task source,
+    //    to reject job’s job promise with a new exception with errorData and a user agent-defined message, in job’s client's Realm.
+    if (job->client) {
+        auto& realm = job->client->realm();
+        HTML::queue_a_task(HTML::Task::Source::DOMManipulation, job->client->responsible_event_loop(), nullptr, JS::create_heap_function(realm.heap(), [&realm, job, message] {
+            HTML::TemporaryExecutionContext const context(*job->client, HTML::TemporaryExecutionContext::CallbacksEnabled::Yes);
+            WebIDL::reject_promise(realm, *job->job_promise, Error::create(realm, message));
+        }));
+    }
+
+    // 2. For each equivalentJob in job’s list of equivalent jobs:
+    for (auto& equivalent_job : job->list_of_equivalent_jobs) {
+        // 1. If equivalentJob’s client is null, continue.
+        if (!equivalent_job->client)
+            continue;
+
+        // 2. Queue a task, on equivalentJob’s client's responsible event loop using the DOM manipulation task source,
+        //    to reject equivalentJob’s job promise with a new exception with errorData and a user agent-defined message,
+        //    in equivalentJob’s client's Realm.
+        auto& realm = equivalent_job->client->realm();
+        HTML::queue_a_task(HTML::Task::Source::DOMManipulation, equivalent_job->client->responsible_event_loop(), nullptr, JS::create_heap_function(realm.heap(), [&realm, equivalent_job, message] {
+            HTML::TemporaryExecutionContext const context(*equivalent_job->client, HTML::TemporaryExecutionContext::CallbacksEnabled::Yes);
+            WebIDL::reject_promise(realm, *equivalent_job->job_promise, Error::create(realm, message));
+        }));
+    }
+}
+
+// https://w3c.github.io/ServiceWorker/#schedule-job-algorithm
 void schedule_job(JS::VM& vm, JS::NonnullGCPtr<Job> job)
 {
     // 1. Let jobQueue be null.
     // Note: See below for how we ensure job queue
 
     // 2. Let jobScope be job’s scope url, serialized.
-    auto job_scope = job->scope_url.serialize();
+    // FIXME: specify whether to use fragment or no in the spec here
```

Comment:
I'm not sure I understand this comment completely, but the spec seems to say that it shouldn't be excluding the fragment